### PR TITLE
Fastdom

### PR DESCRIPTION
**Pull request by gestchild**
References #897

## Type

🚑 Health

## Value
Limits forced reflows and speeds up javascript execution

## Screenshot
Before:
truncate-b4

After:
truncate-after-fastdom

### All
- [X] Demoed to the relevant people
- [X] PR labelled and assigned
- [X] Any introduced code complexity has been flagged

### CODE REVIEW

File: client/js/components/shrink-stories-nav.js [hunk 1]
```diff
@@ -1,42 +1,53 @@
...
   });
-
-  onWindowScrollDebounce$.subscribe({
-    next: () => setIsNarrow(distanceScrolled() > elFromTop)
+  const elHeight = fastdom.measure(() => {
+    return el.offsetHeight;
   });
 
-  dispatch(setStickyNavHeight(el.offsetHeight));
+  Promise.all([elFromTop, elHeight]).then(values => {
```
Comment (jamesgorrie) [inline client/js/components/shrink-stories-nav.js]: I think you could destructure here like:
```JS
Promise.all([elFromTop, elHeight]).then(([fromTop, height]) => {});
```

Just FYI.

File: client/js/components/shrink-stories-nav.js [hunk 2]
```diff
@@ -1,42 +1,53 @@
...
+    const fromTop = values[0];
+    const height = values[1];
+
```
Comment (jamesgorrie) [inline client/js/components/shrink-stories-nav.js]: There's quite a lot of DOM reading and writing in here too? Should we not be wrapping that as well?

File: client/js/components/truncate-text.js [hunk 3]
```diff
@@ -34,25 +37,29 @@ const truncateText = (caption) => {
 
     caption.classList.add(truncateClass);
 
-    if (hasBeenEllipsified(caption)) {
-      caption.parentNode.insertBefore(truncateControl, caption.nextSibling);
-      const truncated$ = fromEvent(truncateControl, 'click').fold((isClosed) => !isClosed, false);
+    hasBeenEllipsified(caption).then((value) => {
+      if (value) {
```
Comment (jamesgorrie) [inline client/js/components/truncate-text.js]: Why would there not be a value? The DOM hasn't loaded?

File: client/js/components/truncate-text.js [hunk 4]
```diff
@@ -34,25 +37,29 @@ const truncateText = (caption) => {
...
     onWindowResizeDebounce$.subscribe({
       next() {
-        if (hasBeenEllipsified(caption)) {
-          truncateControl.classList.remove('is-hidden');
-        } else {
-          truncateControl.classList.add('is-hidden');
-        }
+        hasBeenEllipsified(caption).then((value) => {
+          if (value) {
+            truncateControl.classList.remove('is-hidden');
```
Comment (jamesgorrie) [inline client/js/components/truncate-text.js]: should we not be using `fastdom.mutate` here?

**Comment by jamesgorrie** [comment]
(Sorry for the comments - just saw this was in WIP)

File: client/js/components/truncate-text.js (hunk 3 above)
Comment (gestchild) [inline client/js/components/truncate-text.js]: hasBeenEllipsified returns false if the text isn't truncated. In which case, we don't need the control

File: client/js/util.js [hunk 5]
```diff
@@ -36,11 +36,55 @@ const featureTest = (property, value, noPrefixes) => {
   return mStyle[ property ].indexOf(value) !== -1;
 };
 
+const addClassesToElements = (elements, className) => {
```
Comment (jamesgorrie) [inline client/js/util.js]: Why not embed fastDom here so we don't have to worry about it in the implementation?

**Pull request merged**